prover/strategies/matching.md

```diff
@@ -177,6 +195,279 @@ Recurse over assoc-only constructors (including `pto`):
+                               => P1 ++Patterns P2 ++Patterns Ps
+                  , variables: _
+                  , subst:     _
+                  , rest:      _
+                  )
+
+  // \equals(_,_) mismatched
+  rule #matchAssoc( terms:     T, _
+                  , pattern:   \equals(...), _
+                  , variables: _
+                  , subst:     _
+                  , rest:      _
+                  )
+    => #matchFailure("\\equals(_,_) does not match"), .MatchResults
+    requires \equals(...) :/=K T
+
+  // \not(_) matched
+  rule #matchAssoc( terms:     \not(T), Ts
+                               => T ++Patterns Ts
+                  , pattern:   \not(P), Ps
+                               => P ++Patterns Ps
+                  , variables: _
+                  , subst:     _
+                  , rest:      _
+                  )
+
+  // \not(_) mismatched
+  rule #matchAssoc( terms:     T, _
+                  , pattern:   \not(_), _
+                  , variables: _
+                  , subst:     _
+                  , rest:      _
+                  )
+    => #matchFailure("\\not(_) does not match"), .MatchResults
+    requires \not(...) :/=K T
+
+  // \and(...) matched
+  rule #matchAssoc( terms:     \and(T_ARGS), Ts
+                               => T_ARGS ++Patterns Ts
+                  , pattern:   \and(P_ARGS), Ps
+                               => P_ARGS ++Patterns Ps
+                  , variables: _
+                  , subst:     _
+                  , rest:      _
+                  )
+
+  // \and(...) mismatched
+  rule #matchAssoc( terms:     T, _
+                  , pattern:   \and(_), _
+                  , variables: _
+                  , subst:     _
+                  , rest:      _
+                  )
+    => #matchFailure("\\and(...) does not match"), .MatchResults
+    requires \and(...) :/=K T
+
+  // \or(...) matched
+  rule #matchAssoc( terms:     \or(T_ARGS), Ts
+                               => T_ARGS ++Patterns Ts
+                  , pattern:   \or(P_ARGS), Ps
+                               => P_ARGS ++Patterns Ps
+                  , variables: _
+                  , subst:     _
+                  , rest:      _
+                  )
+
+  // \or(_) mismatched
+  rule #matchAssoc( terms:     T, _
+                  , pattern:   \or(_), _
+                  , variables: _
+                  , subst:     _
+                  , rest:      _
+                  )
+    => #matchFailure("\\or(...) does not match"), .MatchResults
+    requires \or(...) :/=K T
+
+  // \implies(_,_) matched
+  rule #matchAssoc( terms:     \implies(T1, T2), Ts
+                               => T1 ++Patterns T2 ++Patterns Ts
+                  , pattern:   \implies(P1, P2), Ps
+                               => P1 ++Patterns P2 ++Patterns Ps
+                  , variables: _
+                  , subst:     _
+                  , rest:      _
+                  )
```

Comment:
Lets discuss what we want to do with patterns that are not constructors (possibly modulo AC).

e.g. a term `T` to match pattern `A and B`, `T` needs to match both `A` and `B` with some constraints on the substitutions.

What is your use case for all these cases of matching?

Reply:
The use-case is the `rewrite` strategy: given an equality `A=B` (where A,B are patterns), and the goal has shape `C(A')` for some context `C` and pattern `A'` that is matched by `A`, rewrite `C('A)` to `C(B')`.

Reply:
By matching here, I mean syntactical matching: e.g., `\not(\not(A))` is not an instance of `A`.

Reply:
Equality, or rewrite axiom? Do you expect `A` to have quantifiers, negations, \and, \or etc?
I'd rather have a check `isConstructorPattern` before calling the `#match` function, than implementing all these cases in `#match`

Reply:
Would simply causing `#match` to fail for any non-constructor pattern also work?

Reply:
I do not understand this question:
> Equality, or rewrite axiom?

I use axioms (or claims) whose conclusion is an equality
An example is the following axiom `f_0_X`
```
symbol f(Int, Int) : Int
axiom f_0_X: \forall{X{Int}}
  \implies(
    gte(X{Int}, -1),
    \equals(f(0, X{Int}), X{Int})
  )

claim \equals(f(0,1),1)
strategy rewrite -> f_0_X
```
Here I want to pattern match `f(0, X)` against `f(0,1)`. But I also want to support more complicated things, like the following (not tested):
```
axiom A1: \forall x. f(x) = \exists y. y /\ not (y < x)
axiom A2: \forall x. g(f(x))) = x
claim: g(\exists z. z /\ not(z < 3)) = 3
strategy rewrite <- A1 // rewrites goal to g(f(3)) = 3
            . rewrite -> A2 // rewrites to 3 = 3
```

Reply:
> Would simply causing `#match` to fail for any non-constructor pattern also work?

No, that would not work for my use-case.

Reply:
So maybe rename this variant of matching to `#syntacticalMatch`? I want it to work on arbitrary formulas.